GeosCore/diagnostics_mod.F90

```diff
@@ -818,14 +809,115 @@ SUBROUTINE Set_SpcConc_Diags_VVDry( Input_Opt,  State_Chm, State_Diag,     &
+!\\
+!\\
+! !INTERFACE:
+!
+  SUBROUTINE Set_SpcConc_Diags_MND  ( Input_Opt,  State_Chm, State_Diag,     &
+                                      State_Grid, State_Met, RC            )
+!
+! !USES:
+!
+    USE Input_Opt_Mod,  ONLY : OptInput
+    USE State_Met_Mod,  ONLY : MetState
+    USE State_Chm_Mod,  ONLY : ChmState
+    USE State_Diag_Mod, ONLY : DgnMap
+    USE State_Diag_Mod, ONLY : DgnState
+    USE State_Grid_Mod, ONLY : GrdState
+    USE UnitConv_Mod,   ONLY : Convert_Spc_Units
+!
+! !INPUT PARAMETERS:
+!
+    TYPE(OptInput),   INTENT(IN)    :: Input_Opt    ! Input Options object
+    TYPE(GrdState),   INTENT(IN)    :: State_Grid   ! Grid State object
+    TYPE(MetState),   INTENT(IN)    :: State_Met    ! Meteorology State obj
+!
+! !INPUT/OUTPUT PARAMETERS:
+!
+    TYPE(ChmState),   INTENT(INOUT) :: State_Chm    ! Chemistry State object
+    TYPE(DgnState),   INTENT(INOUT) :: State_Diag   ! Diagnsotics State object
+!
+! !OUTPUT PARAMETERS:
+!
+    INTEGER,          INTENT(OUT)   :: RC           ! Success or failure?
+!
+! !REVISION HISTORY:
+!  21 Nov 2021 - H.P. Lin    - Initial version
+!  See https://github.com/geoschem/geos-chem for complete history
+!EOP
+!------------------------------------------------------------------------------
+!BOC
+!
+! !LOCAL VARIABLES:
+!
+    ! Scalars
+    INTEGER               :: N, S
+
+    ! Strings
+    CHARACTER(LEN=255)    :: ErrMsg, ThisLoc, OrigUnit
+
+    ! Objects
+    TYPE(DgnMap), POINTER :: mapData
+
+    !====================================================================
+    ! Set_SpcConc_Diags_MND begins here!
+    !====================================================================
+
+    ! Assume success
+    RC      =  GC_SUCCESS
+    ThisLoc = &
+         ' -> at Set_SpcConc_Diags_MND (in GeosCore/diagnostics_mod.F90)'
+
+    ! Verify that incoming State_Chm%Species units are kg/kg dry air.
+    IF ( TRIM( State_Chm%Spc_Units ) /= 'kg/kg dry' ) THEN
+       ErrMsg = 'Incorrect species units in Set_SpcConc_Diags_MND!'
        CALL GC_Error( ErrMsg, RC, ThisLoc )
        RETURN
     ENDIF
 
-  END SUBROUTINE Set_SpcConc_Diags_VVDry
+    ! Convert to molec/cm3
+    CALL Convert_Spc_Units( Input_Opt, State_Chm, State_Grid, State_Met, &
+                            'molec/cm3', RC, OrigUnit=OrigUnit )
+
+    !=======================================================================
+    ! Copy species to SpeciesConcMND (concentrations diagnostic) [molec/cm3]
+    !=======================================================================
+    IF ( State_Diag%Archive_SpeciesConcMND ) THEN
+
+       ! Point to mapping obj specific to SpeciesConcMND diagnostic collection
+       mapData => State_Diag%Map_SpeciesConcMND
+
+       !$OMP PARALLEL DO       &
+       !$OMP DEFAULT( SHARED ) &
+       !$OMP PRIVATE( N, S   )
+       DO S = 1, mapData%nSlots
+          N = mapData%slot2id(S)
+          State_Diag%SpeciesConcMND(:,:,:,S) = State_Chm%Species(N)%Conc(:,:,:)
```

Comment:
This is where the conversation should be. The equation is the same as used in unitconv_mod for the case of kg/kg -> molec/cm3, which is in subroutine ConvertSpc_KgKgDry_to_MND.

       ! Molecular weight for the species [kg]                                                    
       MW_kg = State_Chm%SpcData(N)%Info%MW_g * 1.e-3_fp

       ! Loop over grid boxes and do the unit conversion                                          
       DO L = 1, State_Grid%NZ
       DO J = 1, State_Grid%NY
       DO I = 1, State_Grid%NX
          State_Chm%Species(N)%Conc(I,J,L) =                             &
                                     State_Chm%Species(N)%Conc(I,J,L)    &
                                     * State_Met%AIRDEN(I,J,L)              &
                                     * ( AVO / MW_kg ) / 1e+6_fp

Reply:
Thanks! This is now fixe in commit [1666335b](https://github.com/geoschem/geos-chem/commit/1666335bda77e786388cc19f214b4d203cbc3117)..